### PR TITLE
UI metrics

### PR DESCRIPTION
**Pull request by divhead**
---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1207919781222044

### CODE REVIEW

File: src/App/App.tsx [hunk 1]
```diff
@@ -73,6 +75,20 @@ function App() {
                 <br />
               </div>
             );
+
+            if (pendingTxn.metricId) {
+              const metricData = metrics.getPendingEvent(pendingTxn.metricId, true);
+              const metricType = metricData?.metricType || "unknownTxn";
+
+              metrics?.sendMetric({
+                event: `${metricType}.failed`,
```
Comment (gmxer) [inline src/App/App.tsx:84]: maybe I'm reading in long order but weird to see "failed" here, in just "metricId" if statement

File: src/context/MetricsContext/errorReporting.ts [hunk 2]
```diff
@@ -1,18 +1,86 @@
...
+export function useUiMetrics() {
+  const { chainId } = useChainId();
+  const fetcher = useOracleKeeperFetcher(chainId);
+  const subaccountAddress = useSubaccountAddress();
+  const isMetamaskMobile = useIsMetamaskMobile();
+
+  const pendingEvents = useRef({});
+  const timers = useRef({});
+
+  const setPendingEvent = useCallback((key: string, eventData: any) => {
```
Comment (gmxer) [inline src/context/MetricsContext/errorReporting.ts:28]: I think eventData should be typed based on metricType, e.x. `MetricEventData = OrderIncreaseMetricEventData | OrderDecreaseMetricEventData | etc`

File: src/components/Synthetics/PositionList/PositionList.tsx [hunk 3]
```diff
@@ -40,6 +43,22 @@ export function PositionList(p: Props) {
   const [, setEditingPositionKey] = usePositionEditorPositionState();
   const isLoading = useIsPositionsLoading();
 
+  useEffect(() => {
+    metrics.startTimer("positionsList");
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (gmxer) [inline src/components/Synthetics/PositionList/PositionList.tsx:48]: let's use `useLatest` instead

File: src/components/Synthetics/TradeBox/hooks/useTradeboxTransactions.tsx [hunk 4]
```diff
@@ -138,9 +217,45 @@ export function useTradeboxTransactions({ setPendingTxns }: TradeboxTransactions
...
+        collateralDeltaAmount: increaseAmounts?.collateralDeltaAmount,
+        swapPath: increaseAmounts?.swapPathStats?.swapPath || [],
+        sizeDeltaUsd: increaseAmounts?.sizeDeltaUsd,
+        sizeDeltaInTokens: increaseAmounts?.sizeDeltaInTokens,
+        triggerPrice: isLimit ? triggerPrice : 0n,
+        acceptablePrice: increaseAmounts?.acceptablePrice,
+        isLong,
+        orderType,
+        executionFee: executionFee?.feeTokenAmount,
+      };
```
Comment (gmxer) [inline src/components/Synthetics/TradeBox/hooks/useTradeboxTransactions.tsx:243]: maybe create some metricData contructor function based on metric type?

right on it seems quite random. Also metricId could be contracted in same function. Returned type could depend on metric type with functino overloading, or at least union type

File: src/components/Synthetics/TradeBox/hooks/useTradeboxTransactions.tsx [hunk 5]
```diff
@@ -96,36 +136,75 @@ export function useTradeboxTransactions({ setPendingTxns }: TradeboxTransactions
...
-      tokensData,
-      toToken,
+      setPendingOrder,
     ]
   );
 
   const onSubmitIncreaseOrder = useCallback(
     function onSubmitIncreaseOrder() {
+      const orderType = isLimit ? OrderType.LimitIncrease : OrderType.MarketIncrease;
+      let metricType: string;
```
Comment (gmxer) [inline src/components/Synthetics/TradeBox/hooks/useTradeboxTransactions.tsx:193]: metricType should have its own type

File: src/components/Synthetics/PositionSeller/PositionSeller.tsx [hunk 6]
```diff
@@ -325,8 +368,29 @@ export function PositionSeller(p: Props) {
...
+      .catch((error) => {
+        metrics.sendMetric({
+          event: `${metricType}.${isUserRejectedActionError(error) ? "rejected" : "fail"}`,
+          isError: true,
+          message: error.message,
+          fields: metrics.getPendingEvent(metricId, true),
+        });
+
+        throw error;
+      });
```
Comment (gmxer) [inline src/components/Synthetics/PositionSeller/PositionSeller.tsx:393]: this logic seems duplicated several times, let's create some factory/promise-wrapper function that would do same thing for all such places

File: src/context/MetricsContext/MetricsContext.tsx [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+import { getAppVersion } from "lib/version";
+import { getWalletNames } from "lib/wallets/getWalletNames";
+import { Context, PropsWithChildren, useMemo, useRef } from "react";
+import { createContext, useContextSelector } from "use-context-selector";
+import mapValues from "lodash/mapValues";
+
+export type MetricsContextType = {
+  sendMetric: (params: {
+    event: string;
+    fields?: Record<string, any>;
```
Comment (gmxer) [inline src/context/MetricsContext/MetricsContext.tsx:14]: metricData?

File: src/context/MetricsContext/MetricsContext.tsx [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+function serializeCustomFields(fields: Record<string, any>) {
+  return mapValues(fields, (v) => {
+    let result;
+
+    if (typeof v === "bigint") {
+      result = v.toString();
+    }
+
+    if (result.length && result.lenth > 150) {
```
Comment (gmxer) [inline src/context/MetricsContext/MetricsContext.tsx:131]: * "lenth" is typo
* I don't see any assignments to `result` variable besides bigint check
* let's do `Array.isArray(result) && result.length > 150` instead

File: src/context/MetricsContext/MetricsContext.tsx [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+      fields?: any;
+      time?: number;
+      isError: boolean;
+      message?: string;
+    }) {
+      const { time, isError, fields, message, event } = params;
+      const wallets = await getWalletNames();
+
+      // eslint-disable-next-line no-console
+      console.log("sendMetric", {
```
Comment (gmxer) [inline src/context/MetricsContext/MetricsContext.tsx:81]: you can use showDebugValues flag here I think?

File: src/context/MetricsContext/MetricsContext.tsx [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+      if (clear) {
+        timers.current[metricId] = undefined;
+      }
+
+      return Date.now() - time;
+    };
+
+    async function sendMetric(params: {
+      event: string;
+      fields?: any;
```
Comment (gmxer) [inline src/context/MetricsContext/MetricsContext.tsx:72]: metricData

File: src/context/MetricsContext/MetricsContext.tsx [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+import { useChainId } from "lib/chains";
+import { getAppVersion } from "lib/version";
+import { getWalletNames } from "lib/wallets/getWalletNames";
+import { Context, PropsWithChildren, useMemo, useRef } from "react";
+import { createContext, useContextSelector } from "use-context-selector";
+import mapValues from "lodash/mapValues";
+
+export type MetricsContextType = {
+  sendMetric: (params: {
+    event: string;
```
Comment (gmxer) [inline src/context/MetricsContext/MetricsContext.tsx:13]: I'm confused but I think it should be either metricType or metricId

File: src/context/MetricsContext/MetricsContext.tsx [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+  }, [fetcher, subaccountAddress]);
+
+  return <context.Provider value={value}>{children}</context.Provider>;
+}
+
+export function useMetricsSelector<Selected>(selector: (s: MetricsContextType) => Selected) {
+  return useContextSelector(context as Context<MetricsContextType>, selector) as Selected;
+}
+
+export function useMetrics() {
```
Comment (gmxer) [inline src/context/MetricsContext/MetricsContext.tsx:119]: No reason to use use-context-selector that way, but I guess it's fine to leave it if we would like to add some selection logic

File: src/context/MetricsContext/MetricsContext.tsx [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+  return useMetricsSelector((s) => s);
+}
+
+function serializeCustomFields(fields: Record<string, any>) {
```
Comment (gmxer) [inline src/context/MetricsContext/MetricsContext.tsx:123]: metricData: MetricData

File: src/context/MetricsContext/errorReporting.ts [hunk 14]
```diff
@@ -1,18 +1,86 @@
...
+    }) {
+      const { time, isError, fields, message, event } = params;
+      const wallets = await getWalletNames();
+
+      const body = {
+        is1ct: Boolean(subaccountAddress),
+        isDev: isDevelopment(),
+        host: window.location.host,
+        url: window.location.href,
+        wallet: wallets.current,
```
Comment (gmxer) [inline src/context/MetricsContext/errorReporting.ts:59]: wallets / walletNames (it's an array of connected extensions names)

File: src/context/SyntheticsEvents/SyntheticsEventsProvider.tsx [hunk 15]
```diff
@@ -138,6 +154,24 @@ export function SyntheticsEventsProvider({ children }: { children: ReactNode })
...
+      if (order) {
+        const metricId = isSwapOrderType(order.orderType)
+          ? getSwapOrderMetricId(order)
+          : getPositionOrderMetricId(order);
+
+        const metricData = metrics.getPendingEvent(metricId, true);
+        const metricType = metricData?.metricType || "unknownOrder";
+
+        metrics.sendMetric({
+          event: `${metricType}.executed`,
```
Comment (gmxer) [inline src/context/SyntheticsEvents/SyntheticsEventsProvider.tsx:168]: sorry, I now see difference between metricType and event

let's make it separate?
```
event: 'executed' | 'failed' | etc
metricType: MetricType
```

File: src/pages/SyntheticsPage/SyntheticsPage.tsx [hunk 16]
```diff
@@ -66,6 +67,8 @@ export function SyntheticsPage(p: Props) {
   const calcSelector = useCalcSelector();
   const [, setPendingTxns] = usePendingTxns();
 
+  const isMobile = useMedia("(max-width: 1100px)");
```
Comment (gmxer) [inline src/pages/SyntheticsPage/SyntheticsPage.tsx:70]: doesn't seem related to the topic but I assume it's agreed with B

File: src/context/SyntheticsEvents/SyntheticsEventsProvider.tsx [hunk 17]
```diff
@@ -101,6 +106,17 @@ export function SyntheticsEventsProvider({ children }: { children: ReactNode })
         return;
       }
 
+      const metricId = isSwapOrderType(data.orderType) ? getSwapOrderMetricId(data) : getPositionOrderMetricId(data);
+      const metricData = metrics.getPendingEvent(metricId);
```
Comment (gmxer) [inline src/context/SyntheticsEvents/SyntheticsEventsProvider.tsx:110]: but it's in fact impossible that metricData is undefined here?

File: src/context/MetricsContext/MetricsContext.tsx [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+  const timers = useRef({});
+
+  const value: MetricsContextType = useMemo(() => {
+    const setPendingEvent = (metricId: string, eventData: any) => {
+      pendingEvents.current[metricId] = { metricId, ...eventData };
+
+      return eventData;
+    };
+
+    const getPendingEvent = (metricId: string, clear?: boolean) => {
```
Comment (gmxer) [inline src/context/MetricsContext/MetricsContext.tsx:42]: getMetricData/setMetricData? not sure if Cache also should be mentioned in name anyhow

**Review by gmxer — CHANGES_REQUESTED** [review]

File: src/App/App.tsx (hunk 1 above)
Comment (divhead) [reply src/App/App.tsx:84]: We send a metric if a metricId was passed in the transaction

**Review by gmxer — APPROVED** [review]

**Review by gmxer — APPROVED** [review]
